### PR TITLE
perf: softmax num stability

### PR DESCRIPTION
**Pull request by nathanielsimard**

### CODE REVIEW

**Review by antimora — CHANGES_REQUESTED** [review]
We need some tests. You can find values here: https://github.com/burn-rs/burn/issues/315

**Comment by nathanielsimard** [comment]
> We need some tests. You can find values here: #315

Softmax and log softmax are tested in burn-tensor and burn-autodiff, if you remove the max it produces nan and the tests don't pass.
Cross entropy is also tested, so I'm confident in the code.
